libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-discovery-feed is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-discovery-feed is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-discovery-feed.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <errno.h>
+#include <math.h>
+#include <string.h>
+
+#include <gio/gio.h>
+
+#include <eos-shard/eos-shard-blob.h>
+#include <eos-shard/eos-shard-shard-file.h>
+#include <eos-shard/eos-shard-record.h>
+
+#include "feed-all-async-tasks-private.h"
+#include "feed-base-card-store.h"
+#include "feed-card-layout-direction.h"
+#include "feed-knowledge-app-artwork-card-store.h"
+#include "feed-knowledge-app-card-store.h"
+#include "feed-knowledge-app-news-card-store.h"
+#include "feed-knowledge-app-video-card-store.h"
+#include "feed-knowledge-app-proxy.h"
+#include "feed-orderable-model.h"
+#include "feed-quote-card-store.h"
+#include "feed-sizes.h"
+#include "feed-store-provider.h"
+#include "feed-text-sanitization.h"
+#include "feed-word-card-store.h"
+#include "feed-word-quote-card-store.h"
+
+typedef GPtrArray * (*ModelsFromResultsAndShardsFunc) (const char * const *shards,
+                                                       GPtrArray          *model_hts,
+                                                       gpointer            user_data);
+typedef GObject * (*ModelFromResultFunc) (GHashTable *model_ht,
+                                          gpointer    user_data);
+
+static GHashTable *
+hashtable_from_variant (GVariant *variant)
```

Comment:
Why do you need to marshal this into an hash table? For instance, you could use `g_variant_lookup()` and pass a reference to the same variant around.

Reply:
After a bit of work I got that working.

However, I'm not super convinced that its a good idea - `g_variant_lookup` allocates its outparam and implements a linear scan. I suppose that converting to a hashtable also allocates, though it moves all that allocation up-front and doesn't force us to litter the code with `g_autofree` everywhere.

For now I've done this change in a fixup commit so we can see how it looks, but if you don't think the tradeoff above is worth it, I'd be pro dropping it